File: src/01-fundamentals/01-exercises/02-immutability/exercise.ts
/**
 * =============================================================================
 * TUTORIAL: INMUTABILIDAD EN PROGRAMACIÓN FUNCIONAL
 * =============================================================================
 * 
 * La inmutabilidad significa que una vez que creas un dato, NUNCA lo modificas.
 * En su lugar, crear NUEVOS datos basados en los originales.
 * 
 * ¿Por qué? Porque los datos mutables causan bugs difíciles de rastrear.
 * =============================================================================
 */

// =============================================================================
// PARTE 1: EL PROBLEMA CON LA MUTABILIDAD
// =============================================================================

/**
 * ❌ MAL: Código mutable - Los datos cambian "debajo de tus pies"
 */
function ejemploMutableProblematico(): void {
    console.log("❌ PROBLEMA CON MUTABILIDAD:");
    
    // Creamos un usuario mutable
    const usuario = {
        nombre: "Ana",
        edad: 25,
        direccion: {
            ciudad: "Madrid",
            calle: "Gran Vía 1"
        }
    };
    
    // Guardamos una "referencia" pensando que es una copia
    const usuarioBackup = usuario;
    
    // Modificamos el usuario original
    usuario.nombre = "María";
    usuario.direccion.ciudad = "Barcelona";
    
    // ¡SORPRESA! El backup también cambió
    console.log("Original:", usuario.nombre);        // María
    console.log("Backup:", usuarioBackup.nombre);    // María (¡debería ser Ana!)
    console.log("¿Son el mismo objeto?", usuario === usuarioBackup); // true
    
    // Esto causa BUGS porque:
    // 1. Esperabas tener dos versiones diferentes
    // 2. El backup era para "revertir" si algo falla
    // 3. Ahora no puedes comparar antes/después
}


// =============================================================================
// PARTE 2: TÉCNICAS DE INMUTABILIDAD EN TYPESCRIPT
// =============================================================================

/**
 * TÉCNICA 1: Usar 'readonly' en tipos
 * 
 * 'readonly' previene modificaciones accidentales a nivel de TypeScript
 * (en runtime JS no hay protección, pero el compilador te avisa)
 * :D
 */

// Definimos un tipo con propiedades readonly :)
type UsuarioInmutable = {
    readonly nombre: string;      // No se puede modificar
    readonly edad: number;        // No se puede modificar
    readonly direccion: {
        readonly ciudad: string;  // Objetos anidados también readonly
        readonly calle: string;
    };
};

// También podemos usar el utility type Readonly<T>
type ProductoInmutable = Readonly<{
    id: string;
    nombre: string;
    precio: number;
}>;

// Para arrays usamos ReadonlyArray<T> o readonly T[]
type ListaInmutable = readonly number[];


/**
 * TÉCNICA 2: Spread Operator para crear copias modificadas
 * 
 * En lugar de mutar, creamos un NUEVO objeto con los cambios
 */
function cambiarNombre(
    usuario: UsuarioInmutable, 
    nuevoNombre: string
): UsuarioInmutable {
    // ❌ ESTO NO COMPILA: usuario.nombre = nuevoNombre;
    // usuario.nombre = nuevoNombre; Cannot assign to 'nombre' because it is a read-only property.ts(2540)
    // usuario.direccion.ciudad = "Barcelona"; Cannot assign to 'ciudad' because it is a read-only property.ts(2540)
    
    // ✅ Creamos un nuevo objeto con el nombre cambiado
    return {
        ...usuario,           // Copiamos todas las propiedades
        nombre: nuevoNombre   // Sobrescribimos solo 'nombre'
    };
}

/**
 * TÉCNICA 3: Copia profunda (deep copy) para objetos anidados
 * 
 * El spread operator hace copia SUPERFICIAL, no profunda.
 * Para objetos anidados, necesitamos cuidado especial.
 */
function cambiarCiudad(
    usuario: UsuarioInmutable,
    nuevaCiudad: string
): UsuarioInmutable {
    return {
        ...usuario,
        direccion: {
            ...usuario.direccion,  // Copiamos el objeto direccion
            ciudad: nuevaCiudad    // Sobrescribimos ciudad
        }
    };
}

/**
 * TÉCNICA 4: Métodos inmutables de arrays
 * 
 * Algunos métodos MUTAN el array (evitar):
 *   - push, pop, shift, unshift
 *   - splice, sort (sin copia), reverse (sin copia)
 * 
 * Otros crean NUEVOS arrays (preferir):
 *   - map, filter, reduce
 *   - concat, slice
 *   - [...array] (spread)
 */

function operacionesInmutablesArray(): void {
    const numeros: readonly number[] = [1, 2, 3, 4, 5];
    
    // ❌ ESTO NO COMPILA con readonly:
    // numeros.push(6);
    // numeros[0] = 100;
    
    // ✅ Crear nuevo array con elemento agregado al final
    const conSeis = [...numeros, 6];
    console.log("Original:", numeros);  // [1, 2, 3, 4, 5]
    console.log("Con 6:", conSeis);     // [1, 2, 3, 4, 5, 6]
    
    // ✅ Crear nuevo array con elemento agregado al inicio
    const conCero = [0, ...numeros];
    console.log("Con 0:", conCero);     // [0, 1, 2, 3, 4, 5]
    
    // ✅ Crear nuevo array sin un elemento (eliminar el 3)
    const sinTres = numeros.filter(n => n !== 3);
    console.log("Sin 3:", sinTres);     // [1, 2, 4, 5]
    
    // ✅ Crear nuevo array con elementos transformados
    const dobles = numeros.map(n => n * 2);
    console.log("Dobles:", dobles);     // [2, 4, 6, 8, 10]
    
    // ✅ Crear nuevo array ordenado (copia primero, luego ordena)
    const desordenados = [3, 1, 4, 1, 5, 9, 2, 6];
    const ordenados = [...desordenados].sort((a, b) => a - b);
    console.log("Original:", desordenados); // [3, 1, 4, 1, 5, 9, 2, 6]
    console.log("Ordenado:", ordenados);    // [1, 1, 2, 3, 4, 5, 6, 9]
}


// =============================================================================
// PARTE 3: PATRONES COMUNES DE INMUTABILIDAD
// =============================================================================

/**
 * PATRÓN 1: Actualizar elemento en array por índice
 */
function actualizarEnIndice<T>(
    array: readonly T[],
    indice: number,
    nuevoValor: T
): T[] {
    // Creamos nuevo array: elementos antes + nuevo valor + elementos después
    return [
        ...array.slice(0, indice),      // Elementos antes del índice
        nuevoValor,                      // El nuevo valor
        ...array.slice(indice + 1)       // Elementos después del índice
    ];
}

// const originalArray = [1, 2, 3, 4, 5];
// const updatedArray = actualizarEnIndice(originalArray, 2, 100);
// console.log(updatedArray); // [1, 2, 100, 4, 5]
// console.log(originalArray); // [1, 2, 3, 4, 5] (sin modificar)

/**
 * PATRÓN 2: Eliminar elemento en array por índice
 */
function eliminarEnIndice<T>(
    array: readonly T[],
    indice: number
): T[] {
    return [
        ...array.slice(0, indice),      // Elementos antes
        ...array.slice(indice + 1)       // Elementos después (saltamos el índice)
    ];
}

/**
 * PATRÓN 3: Agregar elemento en posición específica
 */
function insertarEnIndice<T>(
    array: readonly T[],
    indice: number,
    valor: T
): T[] {
    return [
        ...array.slice(0, indice),      // Elementos antes
        valor,                           // Nuevo elemento
        ...array.slice(indice)           // Elementos desde el índice en adelante
    ];
}

// console.log(insertarEnIndice([1, 2, 3, 4, 5], 2, 100));

/**
 * PATRÓN 4: Actualizar propiedad anidada en objeto
 * 
 * Cuando tienes objetos profundamente anidados, creas copias
 * en cada nivel hasta llegar a la propiedad que quieres cambiar.
 */
type Empresa = {
    readonly nombre: string;
    readonly departamentos: readonly {
        readonly nombre: string;
        readonly empleados: readonly {
            readonly id: number;
            readonly nombre: string;
            readonly salario: number;
        }[];
    }[];
};

function aumentarSalario(
    empresa: Empresa,
    departamentoIdx: number,
    empleadoIdx: number,
    aumento: number
): Empresa {
    return {
        ...empresa,
        departamentos: empresa.departamentos.map((dept, dIdx) => 
            dIdx !== departamentoIdx 
                ? dept  // Departamentos que no cambian
                : {
                    ...dept,
                    empleados: dept.empleados.map((emp, eIdx) =>
                        eIdx !== empleadoIdx
                            ? emp  // Empleados que no cambian
                            : {
                                ...emp,
                                salario: emp.salario + aumento
                            }
                    )
                }
        )
    };
}


// =============================================================================
// PARTE 4: BIBLIOTECAS PARA INMUTABILIDAD
// =============================================================================

/**
 * Para proyectos grandes, considera usar bibliotecas especializadas:
 * 
 * 1. Immer (https://immerjs.github.io/immer/)
 *    - Te permite escribir código "mutable" que produce resultados inmutables
 *    - Muy intuitivo para desarrolladores que vienen de OOP
 * 
 * Ejemplo con Immer:
 * 
 *   import produce from 'immer';
 *   
 *   const nuevoUsuario = produce(usuario, draft => {
 *       draft.direccion.ciudad = "Barcelona";  // ¡Parece mutación pero no lo es!
 *   });
 * 
 * 2. Immutable.js (https://immutable-js.com/)
 *    - Estructuras de datos inmutables optimizadas
 *    - Usa "structural sharing" para eficiencia
 * 
 * 3. fp-ts (https://gcanti.github.io/fp-ts/)
 *    - Biblioteca FP completa para TypeScript
 *    - Incluye Option, Either, y más
 */


// =============================================================================
// PARTE 5: EJERCICIOS PRÁCTICOS
// =============================================================================

/**
 * EJERCICIO 1: Implementa función para agregar item a carrito
 * 
 * El carrito original NO debe modificarse.
 */
type ItemCarrito = {
    readonly productoId: string;
    readonly cantidad: number;
    readonly precio: number;
};

type Carrito = {
    readonly items: readonly ItemCarrito[];
    readonly total: number;
};

// Implementa esta función de forma inmutable:
function agregarAlCarrito(
    carrito: Carrito,
    nuevoItem: ItemCarrito
): Carrito {
    // Tu código aquí
    // Pistas:
    // 1. Crea nuevo array de items con el nuevo item
    // 2. Recalcula el total
    // 3. Retorna un NUEVO carrito
    
    return {
        items: [...carrito.items, nuevoItem],
        total: carrito.total + (nuevoItem.cantidad * nuevoItem.precio)
    };
}

/**
 * EJERCICIO 2: Implementa función para actualizar cantidad
 * 
 * Busca el item por productoId y actualiza su cantidad.
 */
function actualizarCantidad(
    carrito: Carrito,
    productoId: string,
    nuevaCantidad: number
): Carrito {
    // Mapeamos los items, cambiando solo el que coincide con productoId
    const nuevosItems = carrito.items.map(item =>
        item.productoId === productoId
            ? { ...item, cantidad: nuevaCantidad }  // Creamos nuevo item con cantidad actualizada
            : item  // Items que no cambian se mantienen igual
    );
    
    // Recalculamos el total
    const nuevoTotal = nuevosItems.reduce(
        (sum, item) => sum + (item.cantidad * item.precio),
        0
    );
    
    return {
        items: nuevosItems,
        total: nuevoTotal
    };
}

/**
 * EJERCICIO 3: Implementa función para eliminar item del carrito
 */
function eliminarDelCarrito(
    carrito: Carrito,
    productoId: string
): Carrito {
    // Filtramos para excluir el item con el productoId dado
    const nuevosItems = carrito.items.filter(
        item => item.productoId !== productoId
    );
    
    // Recalculamos el total
    const nuevoTotal = nuevosItems.reduce(
        (sum, item) => sum + (item.cantidad * item.precio),
        0
    );
    
    return {
        items: nuevosItems,
        total: nuevoTotal
    };
}


// // =============================================================================
// // DEMOSTRACIÓN
// // =============================================================================

// console.log("=".repeat(70));
// console.log("TUTORIAL: INMUTABILIDAD");
// console.log("=".repeat(70));

// ejemploMutableProblematico();

// console.log("\n✅ OPERACIONES INMUTABLES CON ARRAYS:");
// operacionesInmutablesArray();

// console.log("\n✅ EJERCICIOS DEL CARRITO:");

// const carritoVacio: Carrito = { items: [], total: 0 };

// const conPrimerItem = agregarAlCarrito(carritoVacio, {
//     productoId: "A1",
//     cantidad: 2,
//     precio: 100
// });
// console.log("Carrito vacío:", carritoVacio);
// console.log("Con primer item:", conPrimerItem);

// const conSegundoItem = agregarAlCarrito(conPrimerItem, {
//     productoId: "B2",
//     cantidad: 1,
//     precio: 50
// });
// console.log("Con segundo item:", conSegundoItem);

// const cantidadActualizada = actualizarCantidad(conSegundoItem, "A1", 5);
// console.log("Cantidad actualizada:", cantidadActualizada);

// const itemEliminado = eliminarDelCarrito(cantidadActualizada, "A1");
// console.log("Item eliminado:", itemEliminado);

// console.log("\n" + "=".repeat(70));
// console.log("BENEFICIOS DE INMUTABILIDAD:");
// console.log("- Historial completo: todos los estados anteriores existen");
// console.log("- Debugging fácil: puedes comparar antes/después");
// console.log("- Thread-safe: no hay race conditions");
// console.log("- Predecible: los datos nunca cambian inesperadamente");
// console.log("=".repeat(70));


export {};

File: src/01-fundamentals/01-exercises/02-immutability/Exercise.scala
/**
 * =============================================================================
 * TUTORIAL: INMUTABILIDAD EN SCALA
 * =============================================================================
 * 
 * Scala está diseñado para la inmutabilidad desde su núcleo.
 * - 'val' en lugar de 'var'
 * - 'case class' son inmutables por defecto
 * - Colecciones inmutables son el default
 * 
 * =============================================================================
 */

object ImmutabilityTutorial extends App {

  // ===========================================================================
  // PARTE 1: val vs var
  // ===========================================================================

  /**
   * 'val' = valor inmutable (preferir siempre)
   * 'var' = variable mutable (evitar)
   */
  
  val nombre: String = "Ana"    // Inmutable - no puede cambiar
  // nombre = "María"           // ❌ Error de compilación!
  
  var contador: Int = 0         // Mutable - puede cambiar
  contador = 1                  // ✅ Compila (pero evitar)


  // ===========================================================================
  // PARTE 2: case class - Objetos Inmutables
  // ===========================================================================

  /**
   * case class crea automáticamente:
   * - Constructor con todos los campos
   * - Método copy() para crear copias modificadas
   * - equals() y hashCode() basados en valores
   * - toString() legible
   */
  case class Usuario(
    nombre: String,    // Inmutable por defecto
    edad: Int,
    email: String
  )

  // Creamos un usuario
  val usuario1 = Usuario("Ana", 25, "ana@email.com")
  
  // No podemos modificarlo directamente:
  // usuario1.edad = 26  // ❌ Error de compilación!
  
  // Para "modificar", creamos una COPIA con el cambio:
  val usuario2 = usuario1.copy(edad = 26)
  
  println(s"Usuario original: $usuario1")  // Usuario(Ana,25,ana@email.com)
  println(s"Usuario copia: $usuario2")      // Usuario(Ana,26,ana@email.com)
  println(s"¿Son iguales? ${usuario1 == usuario2}")  // false (por el contenido)


  // ===========================================================================
  // PARTE 3: Objetos Anidados
  // ===========================================================================

  /**
   * Cuando tienes case classes anidados, usas copy() en cada nivel
   */
  case class Direccion(ciudad: String, calle: String)
  
  case class Persona(
    nombre: String,
    direccion: Direccion
  )

  val persona = Persona("Carlos", Direccion("Madrid", "Gran Vía 1"))
  
  // Para cambiar la ciudad, necesitamos copy() anidado:
  val personaEnBarcelona = persona.copy(
    direccion = persona.direccion.copy(ciudad = "Barcelona")
  )
  
  println(s"\nPersona original: $persona")
  println(s"Persona en Barcelona: $personaEnBarcelona")


  // ===========================================================================
  // PARTE 4: Colecciones Inmutables
  // ===========================================================================

  /**
   * En Scala, las colecciones por defecto son INMUTABLES:
   * - List, Vector, Set, Map
   * 
   * Cuando "agregas" un elemento, obtienes una NUEVA colección.
   * La original no cambia.
   */
  
  val numeros: List[Int] = List(1, 2, 3, 4, 5)
  
  // Agregar al inicio (operador ::)
  // El :: "prepend" crea una nueva lista con el elemento al inicio
  val conCero: List[Int] = 0 :: numeros
  println(s"\nOriginal: $numeros")       // List(1, 2, 3, 4, 5)
  println(s"Con 0 al inicio: $conCero")  // List(0, 1, 2, 3, 4, 5)
  
  // Agregar al final (operador :+)
  val conSeis: List[Int] = numeros :+ 6
  println(s"Con 6 al final: $conSeis")   // List(1, 2, 3, 4, 5, 6)
  
  // Concatenar listas (operador ++)
  val combinada: List[Int] = numeros ++ List(6, 7, 8)
  println(s"Combinada: $combinada")      // List(1, 2, 3, 4, 5, 6, 7, 8)
  
  // Filtrar elementos
  val pares: List[Int] = numeros.filter(_ % 2 == 0)
  println(s"Solo pares: $pares")         // List(2, 4)
  
  // Transformar elementos
  val dobles: List[Int] = numeros.map(_ * 2)
  println(s"Dobles: $dobles")            // List(2, 4, 6, 8, 10)
  
  // ¡La lista original NUNCA cambió!
  println(s"Original (sin cambios): $numeros")  // List(1, 2, 3, 4, 5)


  // ===========================================================================
  // PARTE 5: Actualizar Elementos en Colecciones
  // ===========================================================================

  /**
   * Para "actualizar" un elemento, creamos una nueva colección
   * con el elemento modificado.
   */
  
  // Actualizar elemento en posición específica
  def actualizarEnIndice[A](lista: List[A], indice: Int, nuevoValor: A): List[A] = {
    // zipWithIndex añade el índice a cada elemento: List((a,0), (b,1), ...)
    // map transforma cada elemento
    lista.zipWithIndex.map {
      case (_, i) if i == indice => nuevoValor  // Si es el índice, usamos nuevo valor
      case (valor, _) => valor                   // Si no, mantenemos el original
    }
  }
  
  val letras = List("a", "b", "c", "d")
  val letrasActualizadas = actualizarEnIndice(letras, 1, "X")
  println(s"\nLetras originales: $letras")         // List(a, b, c, d)
  println(s"Letras actualizadas: $letrasActualizadas")  // List(a, X, c, d)

  // También puedes usar updated() en Vector (más eficiente para acceso por índice)
  val vector = Vector("a", "b", "c", "d")
  val vectorActualizado = vector.updated(1, "X")
  println(s"Vector actualizado: $vectorActualizado")  // Vector(a, X, c, d)


  // ===========================================================================
  // PARTE 6: Mapas Inmutables
  // ===========================================================================

  /**
   * Los Map en Scala también son inmutables por defecto.
   */
  
  val precios: Map[String, Double] = Map(
    "manzana" -> 1.50,
    "naranja" -> 2.00,
    "plátano" -> 1.00
  )
  
  // Agregar/actualizar: crea nuevo Map
  val preciosActualizados = precios + ("pera" -> 1.75)
  println(s"\nPrecios originales: $precios")
  println(s"Precios con pera: $preciosActualizados")
  
  // Eliminar: crea nuevo Map
  val sinNaranja = preciosActualizados - "naranja"
  println(s"Sin naranja: $sinNaranja")
  
  // El Map original no cambió
  println(s"Original sin cambios: $precios")


  // ===========================================================================
  // PARTE 7: Patrón Lens (Lentes) para Actualizaciones Anidadas
  // ===========================================================================

  /**
   * Para estructuras muy anidadas, el patrón copy() se vuelve tedioso.
   * La solución son los "Lenses" - una técnica FP avanzada.
   * 
   * Bibliotecas como Monocle (https://www.optics.dev/Monocle/) proveen lenses.
   * 
   * Ejemplo conceptual (sin biblioteca):
   */
  
  case class Calle(nombre: String, numero: Int)
  case class DireccionCompleta(calle: Calle, ciudad: String, pais: String)
  case class Empleado(nombre: String, direccion: DireccionCompleta)
  case class Empresa(nombre: String, empleados: List[Empleado])

  val miEmpresa = Empresa(
    "TechCorp",
    List(
      Empleado("Ana", DireccionCompleta(Calle("Gran Vía", 10), "Madrid", "España")),
      Empleado("Bob", DireccionCompleta(Calle("Main St", 100), "NYC", "USA"))
    )
  )

  // Cambiar el número de calle del primer empleado (¡muy anidado!)
  val empresaActualizada = miEmpresa.copy(
    empleados = miEmpresa.empleados.zipWithIndex.map {
      case (emp, 0) => emp.copy(
        direccion = emp.direccion.copy(
          calle = emp.direccion.calle.copy(numero = 20)
        )
      )
      case (emp, _) => emp
    }
  )

  println(s"\n Empresa original primer empleado: ${miEmpresa.empleados.head.direccion.calle}")
  println(s"Empresa actualizada primer empleado: ${empresaActualizada.empleados.head.direccion.calle}")


  // ===========================================================================
  // PARTE 8: EJERCICIO PRÁCTICO - Carrito de Compras
  // ===========================================================================

  case class ItemCarrito(productoId: String, cantidad: Int, precio: Double)
  case class Carrito(items: List[ItemCarrito], total: Double)

  // Función para agregar item al carrito (inmutable)
  def agregarAlCarrito(carrito: Carrito, nuevoItem: ItemCarrito): Carrito = {
    // Creamos nueva lista de items con el nuevo item al final
    val nuevosItems = carrito.items :+ nuevoItem
    // Calculamos nuevo total
    val nuevoTotal = carrito.total + (nuevoItem.cantidad * nuevoItem.precio)
    // Retornamos NUEVO carrito
    Carrito(nuevosItems, nuevoTotal)
  }

  // Función para actualizar cantidad (inmutable)
  def actualizarCantidad(carrito: Carrito, productoId: String, nuevaCantidad: Int): Carrito = {
    val nuevosItems = carrito.items.map { item =>
      if (item.productoId == productoId) item.copy(cantidad = nuevaCantidad)
      else item
    }
    val nuevoTotal = nuevosItems.map(i => i.cantidad * i.precio).sum
    Carrito(nuevosItems, nuevoTotal)
  }

  // Función para eliminar item (inmutable)
  def eliminarDelCarrito(carrito: Carrito, productoId: String): Carrito = {
    val nuevosItems = carrito.items.filter(_.productoId != productoId)
    val nuevoTotal = nuevosItems.map(i => i.cantidad * i.precio).sum
    Carrito(nuevosItems, nuevoTotal)
  }

  // Demostración
  println("\n" + "=" * 70)
  println("EJERCICIO: CARRITO INMUTABLE")
  println("=" * 70)

  val carritoVacio = Carrito(List.empty, 0.0)
  println(s"Carrito vacío: $carritoVacio")

  val conItem1 = agregarAlCarrito(carritoVacio, ItemCarrito("A1", 2, 100.0))
  println(s"Con item 1: $conItem1")

  val conItem2 = agregarAlCarrito(conItem1, ItemCarrito("B2", 1, 50.0))
  println(s"Con item 2: $conItem2")

  val cantidadActualizada = actualizarCantidad(conItem2, "A1", 5)
  println(s"Cantidad actualizada: $cantidadActualizada")

  val itemEliminado = eliminarDelCarrito(cantidadActualizada, "A1")
  println(s"Item eliminado: $itemEliminado")

  // ¡Todos los carritos anteriores siguen existiendo!
  println(s"\nCarrito original sigue igual: $conItem2")

  println("\n" + "=" * 70)
  println("RESUMEN: En Scala, la inmutabilidad es el DEFAULT")
  println("- Usa 'val' en lugar de 'var'")
  println("- Usa 'case class' para objetos de datos")
  println("- Usa colecciones inmutables (List, Vector, Map, Set)")
  println("- Usa copy() para crear copias modificadas")
  println("=" * 70)
}
